### ISSUE TITLE
After we update to version 2.8.1, ES module Error happens 

### ISSUE THREAD

**Issue by SeonyuK** [carbon-design-system/ibm-products#3446]
## What package(s) are you using?

- [X] Carbon for IBM Products (or Carbon for IBM Cloud & Cognitive)
- [ ] Legacy/CDAI
- [ ] Legacy/Security

## Detailed description

> Describe in detail the issue you're having.
After updating to version 2.8.0 this error happens
```
Error: require() of ES Module /Users/jeremy/Workspace/ui-management/node_modules/react-dnd/dist/index.js from /Users/jeremy/Workspace/ui-management/node_modules/@carbon/ibm-products/lib/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js not supported.
Instead change the require of index.js in /Users/jeremy/Workspace/ui-management/node_modules/@carbon/ibm-products/lib/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js to a dynamic import() which is available in all CommonJS modules.

```
> Is this issue related to a specific component?
no
> What did you expect to happen? What happened instead? What would you like to
> see changed?
not having the error
> What browser are you working in?

> What version of the @carbon/ibm-products (or @carbon/ibm-cloud-cognitive)
2.8.0
> package are you using?

> What offering/product do you work on? Any pressing ship or release dates we
> should be aware of?

## Steps to reproduce the issue

1. Step one
2. Step two
3. Step three
4. etc.

> Please create a reduced test case in CodeSandbox
>
> https://codesandbox.io/s/github/carbon-design-system/ibm-products/tree/main/examples/carbon-for-ibm-products/SidePanel

## Additional information

- Screenshots or code
![Screenshot 2023-09-06 at 12 04 18 PM](https://github.com/carbon-design-system/ibm-products/assets/144151347/5bf58523-fb76-44f4-abe8-195146209d96)

- Notes

**Comment by elycheea** [comment]
So it looks like the latest update of [`react-dnd`](https://github.com/react-dnd/react-dnd/releases/tag/v16.0.0) is now pure ESM. I do think we need to be on v16, which is why we update the package when we moved to React 18.

It looks like a fix to continue CJS support was merged but never published, and it’s been some time.

Without this fix, we may need to evaluate alternatives, which could push our stable releases back, or consider dropping support for CJS. 😞

**Comment by lee-chase** [comment]
Hi @SeonyuK can you confirm your build/setup? Are. you using Webpack, Vite, NextJS or other? An example of this being broken would really help.

It may just be that we need to find the right configuration to correctly build with react-dnd

**Comment by vlukashov** [comment]
We've run into the same issue on a NextJS project (different team, not the one @SeonyuK is on). Here is a repro case: https://stackblitz.com/edit/next-typescript-kodtrr

As a workaround we've ended up dynamically importing the `@carbon/ibm-products` package. Apparently that happens to trigger NextJS onto the ESM import path, whereas for static imports of `@carbon/ibm-products` it uses the CJS import path.

**Comment by SeonyuK** [comment]
We use  NextJS.

**Comment by lee-chase** [comment]
@vlukashov @SeonyuK I'm trying to find a way to avoid switching dnd libraries.

Would really like to hear your thoughts on the impact of dropping the CommonJS build. 

This issue here suggests a workaround. 

https://github.com/udecode/plate/issues/1609

```
const nextConfig = {
  experimental: {
    transpilePackages: [
      '@react-dnd/asap',
      '@react-dnd/invariant',
      '@react-dnd/shallowequal',
      'dnd-core',
      'react-dnd',
      'react-dnd-html5-backend',
    ],
  },
},
```

**Comment by lee-chase** [comment]
Do either of you have a simple NextJS example, that demonstrates the issue. It would be helpful to ensure any workarounds or fixes meet your needs.

**Comment by vlukashov** [comment]
> This issue here suggests a workaround.
> 
> [udecode/plate#1609](https://github.com/udecode/plate/issues/1609)
> 
> ```
> const nextConfig = {
>   experimental: {
>     transpilePackages: [
>       '@react-dnd/asap',
>       '@react-dnd/invariant',
>       '@react-dnd/shallowequal',
>       'dnd-core',
>       'react-dnd',
>       'react-dnd-html5-backend',
>     ],
>   },
> },
> ```

Thanks for researching this issue! We are on NextJS v13, but sadly the workaround did not work for me–at least not "as-is". I have not spent more time digging into the inner workings of `transpilePackages` to see if some variation of it could work.

> Would really like to hear your thoughts on the impact of dropping the CommonJS build.

That would resolve the issue for us 👍. After all, our current workaround is to force NextJS into using the ESM build of `@carbon/ibm-products`.

**Comment by lee-chase** [comment]
So I created a NextJS example on CodeSandbox and v2.9, what do I need to do to make it fail? https://codesandbox.io/p/github/lee-chase/next-c4p/main?file=/package.json:1,1&workspaceId=7d09c446-5d37-4748-92d5-3d510c6572d4

**Comment by vlukashov** [comment]
> perhaps you could test that theory by deleting the lib folder from node_modules and removing this line from our package.json"main": "lib/index.js",

Checks out. After I manually deleted the `node_modules/@carbon/ibm-products/lib` folder, and the `"main": "lib/index.js",` line from `node_modules/@carbon/ibm-products/package.json`, NextJS is happy to import it as ESM without using any workarounds.

**Comment by SeonyuK** [comment]
> So I created a NextJS example on CodeSandbox and v2.9, what do I need to do to make it fail? https://codesandbox.io/p/github/lee-chase/next-c4p/main?file=/package.json:1,1&workspaceId=7d09c446-5d37-4748-92d5-3d510c6572d4

I can't access this link and it happened v2.8. I checked with v2.9 and it is working fine

**Comment by lee-chase** [comment]
@SeonyuK try [this link](https://codesandbox.io/p/github/lee-chase/next-c4p/main?file=%2Fsrc%2Fcomponents%2FExample.js%3A28%2C25&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clmqgqx7s000i3b6krknmh1zt%2522%252C%2522sizes%2522%253A%255B44.13793103448276%252C55.86206896551724%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clmqgqx7s000e3b6kpa302b8w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clmqgqx7s000g3b6kd9im659n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clmqgqx7s000h3b6kir57qt1e%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clmqgqx7s000e3b6kpa302b8w%2522%253A%257B%2522id%2522%253A%2522clmqgqx7s000e3b6kpa302b8w%2522%252C%2522activeTabId%2522%253A%2522clmrrps5q01n23b6pdlxktfoy%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponents%252FExample.js%2522%252C%2522id%2522%253A%2522clmrrps5q01n23b6pdlxktfoy%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522clmqgqx7s000h3b6kir57qt1e%2522%253A%257B%2522id%2522%253A%2522clmqgqx7s000h3b6kir57qt1e%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clmqgrnmz00l33b6k95sr2fje%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clmrp17gr00nm3b6pirmdddeu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clmqgrnmz00l33b6k95sr2fje%2522%257D%252C%2522clmqgqx7s000g3b6kd9im659n%2522%253A%257B%2522id%2522%253A%2522clmqgqx7s000g3b6kd9im659n%2522%252C%2522activeTabId%2522%253A%2522clmrrpecv016j3b6px66eyeml%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clmqgqx7s000f3b6kpl1ffg3a%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clmrrp9ni0004e5i6dxbaakv9%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clmrrpecv016j3b6px66eyeml%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I do not disagree with the sentiment @SeonyuK that our dependencies should not impact you, however it looks like this very popular package is now in a state of limbo. The realistic options we have are:

1. Wait for an update, but the package may be unmaintained at this point.
2. Take a copy of the already fixed source, then compile it as part of our source.
3. Find a different package, which may have other consequences.
4. Write our own Drag and drop functionality, which increases maintenance.
5. Abandon CommonJS, or find a way to default to ESM as @vlukashov. Which would be a breaking change that we are not quite ready for just yet.

**Comment by lee-chase** [comment]
Created this V12 version - which works (ibm-products 2.7.0) but has a lot of complaints about useLayoutEffect, probably my lack of NextJS understanding (issue https://github.com/lee-chase/next-v12-c4p/issues/1)
- https://github.com/lee-chase/next-v12-c4p

Created this V13 example https://github.com/lee-chase/next-c4p (ibm-products 2.8.1) which works if I edit `node_modules/@carbon/ibm-products/package.json` and switch these to have `module` first.

```
  "module": "es/index.js",
  "main": "lib/index.js",
```

**Comment by elycheea** [comment]
This should be resolved in next release with the merge of #3513
